### ISSUE TITLE
evaluate_function's bookkeeping is wrong for constraints

### ISSUE THREAD

**Issue by brockho** [numbbo/coco#1393]
In the case of constraints, no restarts are triggered at the moment because the bookkeeping of the best observed `f`-value in the `evaluate_function` call does not take into account whether the evaluated point is actually feasible:
```
/* A little bit of bookkeeping */
  if (y[0] < problem->best_observed_fvalue[0]) {
    problem->best_observed_fvalue[0] = y[0];
    problem->best_observed_evaluation[0] = problem->evaluations;
  }
```

This should be fixed to get reasonable results on the `bbob-constrained` test suite.

**Comment by brockho** [comment]
After the discussion with @nikohansen, @asmaatamna, @anneauger, and @dtusar, it seems as if we can simply evaluate the constraints within the above if before to update the `best_observed_fvalue` in case of a constrained problem and only do the update if the solution is feasible.

**Comment by brockho** [comment]
This issue should be solved with the pull request #1406 now.
